widget/slider.go

```diff
@@ -0,0 +1,289 @@
+	thumb := &canvas.Circle{
+		FillColor:   theme.TextColor(),
+		StrokeWidth: 0}
+
+	objects := []fyne.CanvasObject{track, active, thumb}
+
+	return &sliderRenderer{track, active, thumb, objects, s}
+}
+
+type sliderRenderer struct {
+	track  *canvas.Rectangle
+	active *canvas.Rectangle
+	thumb  *canvas.Circle
+
+	objects []fyne.CanvasObject
+	slider  *Slider
+}
+
+// ApplyTheme is called when the Slider may need to update its look
+func (s *sliderRenderer) ApplyTheme() {
+	s.track.FillColor = theme.ButtonColor()
+	s.thumb.FillColor = theme.TextColor()
+	s.active.FillColor = theme.TextColor()
+	s.Refresh()
+}
+
+// Refresh is used to update the widget state for drawing
+func (s *sliderRenderer) Refresh() {
+	s.Layout(s.slider.Size())
+	canvas.Refresh(s.slider)
+}
+
+// Layout the components of the slider widget
+func (s *sliderRenderer) Layout(size fyne.Size) {
+	padLen := theme.Padding()
+	sideLen := padLen * standardScale
+	activeOffset, thumbOffset := s.moveSlide(sideLen)
+
+	var trackPos, activePos, thumbPos fyne.Position
+	var trackSize, activeSize fyne.Size
+
+	if s.slider.Vertical {
+		trackPos = fyne.NewPos(size.Width/2, 0)
+		activePos = fyne.NewPos(trackPos.X, activeOffset)
+
+		trackSize = fyne.NewSize(padLen, size.Height)
+		activeSize = fyne.NewSize(padLen, trackSize.Height-activeOffset)
+
+		thumbPos = fyne.NewPos(
+			trackPos.X-(sideLen-trackSize.Width)/2, thumbOffset)
+	} else {
+		trackPos = fyne.NewPos(0, size.Height/2)
+		activePos = trackPos
+
+		trackSize = fyne.NewSize(size.Width, padLen)
+		activeSize = fyne.NewSize(activeOffset, padLen)
+
+		thumbPos = fyne.NewPos(
+			thumbOffset, trackPos.Y-(sideLen-trackSize.Height)/2)
+	}
+
+	s.track.Move(trackPos)
+	s.track.Resize(trackSize)
+
+	s.active.Move(activePos)
+	s.active.Resize(activeSize)
+
+	s.thumb.Move(thumbPos)
+	s.thumb.Resize(fyne.NewSize(sideLen, sideLen))
+}
+
+// MinSize calculates the minimum size of a slider widget
+func (s *sliderRenderer) MinSize() fyne.Size {
+	s1, s2 := minLongSide, theme.Padding()*standardScale
+	if s.slider.Vertical {
+		return fyne.NewSize(s2, s1)
+	}
+	return fyne.NewSize(s1, s2)
+}
+
+func (s *sliderRenderer) BackgroundColor() color.Color {
+	return theme.BackgroundColor()
+}
+
+func (s *sliderRenderer) Destroy() {
+}
+
+func (s *sliderRenderer) Objects() []fyne.CanvasObject {
+	return s.objects
+}
+
+func (s *sliderRenderer) moveSlide(diameter int) (int, int) {
+	w := s.slider
+	t := s.track.Size()
+	ratio := (w.Value - w.Min) / (w.Max - w.Min)
+	if w.Vertical {
+		y := float64(t.Height) - (ratio * float64(t.Height))
+		return int(y), int(y) - int((1-ratio)*float64(diameter))
+	}
+	x := ratio * float64(t.Width)
+	return int(x), int(x) - int(ratio*float64(diameter))
+}
+
+func checkStep(s, max, min float64) {
+	// make sure there is a positive step and it is less than
+	// the maximum value
+	if s <= 0 {
+		fyne.LogError("Step is less than or equal to zero.", nil)
+	}
+	if s*s >= (max-min)*(max-min) {
```

Comment:
Should this not just be `if s > max-min`?
And surely an equal amount is allowed, otherwise [0, 1] range with step 1 would not be possible?

Reply:
You are right about the equals sign. As far as the squaring is concerned, its due to some extreme paranoia and the idea that max-min is a distance which should never be negative.

Reply:
Fair enough about the paranoia - but math.Abs would make this clearer...

Reply:
I supposed the way the ideas are constructed in my head and their relationship with the code it is clearer to myself. As to what is or is not clearer to someone else reading it I haven't the faintest if idea, but I suppose from a performance standpoint that flipping a single bit would be faster than two multiplication operations.